addons/survey/tests/test_survey.py

```diff
@@ -691,7 +698,87 @@ def test_notify_subscribers(self):
-        for status, count in [(total['text'], total['count']) for total in questions_statistics['totals']]:
-            self.assertEqual(count, 1 if status == expected_answer_status else 0)
+    def test_survey_session_speed_reward_config_propagation(self):
+        """Test the speed rating time limit propagation to non time-customized questions."""
+        test_survey = self.env['survey.survey'].create({
+            'title': 'Test This Survey',
+            'scoring_type': 'scoring_with_answers',
+            'question_and_page_ids': [
+                Command.create({
+                    'is_time_customized': True,
+                    'is_time_limited': True,
+                    'title': 'Question A',
+                }), Command.create({
+                    'is_time_customized': True,
+                    'is_time_limited': True,
+                    'time_limit': 40,
+                    'title': 'Question B',
+                }), Command.create({'title': 'Question C'}),
+            ],
+        })
+        self.assertFalse(test_survey.session_speed_rating)
+
+        test_survey.session_speed_rating = True
+        self.assertEqual(test_survey.session_speed_rating_time_limit, 30)
+        self.assertSetEqual({*test_survey.question_ids.mapped('is_time_limited')}, {True})
+        self.assertListEqual(test_survey.question_ids.mapped('time_limit'), [30, 40, 30])
+        self.assertListEqual(test_survey.question_ids.mapped('is_time_customized'), [False, True, False])
+
+        test_survey.session_speed_rating_time_limit = 40
+        self.assertSetEqual({*test_survey.question_ids.mapped('time_limit')}, {40})
+        self.assertSetEqual({*test_survey.question_ids.mapped('is_time_customized')}, {False})
+
+        test_survey.question_ids[:2].write({
+            "is_time_limited": False,
+            'is_time_customized': True,  # As would do the client
+        })
+        self.assertListEqual(test_survey.question_ids.mapped('is_time_limited'), [False, False, True])
+        self.assertListEqual(test_survey.question_ids.mapped('is_time_customized'), [True, True, False])
+
+        test_survey.session_speed_rating = False
+        self.assertSetEqual({*test_survey.question_ids.mapped('is_time_limited')}, {False})
+        self.assertSetEqual({*test_survey.question_ids.mapped('is_time_customized')}, {False})
+
+    def test_survey_time_limits_results(self):
+        start_time = datetime.datetime(2023, 7, 7, 12, 0, 0)
+        test_survey = self.env['survey.survey'].create({
+            'title': 'Test This Survey',
+            'scoring_type': 'scoring_with_answers',
+            'scoring_success_min': 80.0,
+            'session_speed_rating': True,
+            'session_speed_rating_time_limit': 30,
+            'session_question_start_time': start_time,
+        })
+        answer_correct, answer_incorrect = self.env['survey.question.answer'].create([
```

Comment:
Can't you make embedded create (and find the right one using filtered / order of values) ? My point is that we should probably make question_id required at some point so it should become a o2m command (and don't use Command, it is depreated).